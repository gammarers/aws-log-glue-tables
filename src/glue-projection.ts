
export namespace GlueProjection {

  export interface TimestampRange {
    readonly start: string;
    readonly end?: string;
  }

  export interface Timestamp {
    readonly range: TimestampRange;
  }

}