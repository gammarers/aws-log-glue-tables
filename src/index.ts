export * from './tables/bucket-static-web-hosting-access-log-table';
export * from './tables/cloudfront-access-log-table';
export * from './glue-projection';
