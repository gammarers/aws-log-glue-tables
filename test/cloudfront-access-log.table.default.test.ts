import { App, Stack } from 'aws-cdk-lib';
import { Template } from 'aws-cdk-lib/assertions';
import * as glue from 'aws-cdk-lib/aws-glue';
import * as s3 from 'aws-cdk-lib/aws-s3';
import { CloudFrontAccessLogTable } from '../src';

describe('CloudFrontAccessLogTable default Testing', () => {

  const app = new App();
  const stack = new Stack(app, 'TestingStack');

  const bucket = new s3.Bucket(stack, 'LogBucket');

  const databaseName = 'example-database';

  const tableName = 'example-cloudfront-access-log-table';

  const table = new CloudFrontAccessLogTable(stack, 'CloudFrontAccessLogTable', {
    logBucket: bucket,
    glueDatabaseName: databaseName,
    glueTableName: tableName,
    location: 'path/to/location',
    timestamp: {
      range: {
        start: '2024/11/01',
      },
    },
  });

  it('Is Glue Table', () => {
    expect(table.glueTable).toBeInstanceOf(glue.CfnTable);
  });

  const template = Template.fromStack(stack);

  it('Should match snapshot', () => {
    expect(template.toJSON()).toMatchSnapshot();
  });
});