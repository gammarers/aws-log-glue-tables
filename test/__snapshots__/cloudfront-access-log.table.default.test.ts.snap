// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CloudFrontAccessLogTable default Testing Should match snapshot 1`] = `
{
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "CloudFrontAccessLogTableCloudFrontAccessLogGlueTable613BEDF1": {
      "Properties": {
        "CatalogId": {
          "Ref": "AWS::AccountId",
        },
        "DatabaseName": "example-database",
        "TableInput": {
          "Name": "example-cloudfront-access-log-table",
          "Parameters": {
            "classification": "cloudfront",
            "projection.enabled": "true",
            "projection.timestamp.format": "yyyy/MM/dd",
            "projection.timestamp.range": "2024/11/01,NOW",
            "projection.timestamp.type": "date",
            "skip.header.line.count": "2",
            "storage.location.template": {
              "Fn::Join": [
                "",
                [
                  "s3://",
                  {
                    "Ref": "LogBucketCC3B17E8",
                  },
                  "/path/to/location/\${timestamp}/",
                ],
              ],
            },
          },
          "PartitionKeys": [
            {
              "Name": "timestamp",
              "Type": "string",
            },
          ],
          "StorageDescriptor": {
            "Columns": [
              {
                "Name": "date",
                "Type": "date",
              },
              {
                "Name": "time",
                "Type": "string",
              },
              {
                "Name": "x_edge_location",
                "Type": "string",
              },
              {
                "Name": "sc_bytes",
                "Type": "bigint",
              },
              {
                "Name": "c_ip",
                "Type": "string",
              },
              {
                "Name": "cs_method",
                "Type": "string",
              },
              {
                "Name": "cs_host",
                "Type": "string",
              },
              {
                "Name": "cs_uri_stem",
                "Type": "string",
              },
              {
                "Name": "sc_status",
                "Type": "int",
              },
              {
                "Name": "cs_referrer",
                "Type": "string",
              },
              {
                "Name": "cs_user_agent",
                "Type": "string",
              },
              {
                "Name": "cs_uri_query",
                "Type": "string",
              },
              {
                "Name": "cs_cookie",
                "Type": "string",
              },
              {
                "Name": "x_edge_result_type",
                "Type": "string",
              },
              {
                "Name": "x_edge_request_id",
                "Type": "string",
              },
              {
                "Name": "x_host_header",
                "Type": "string",
              },
              {
                "Name": "cs_protocol",
                "Type": "string",
              },
              {
                "Name": "cs_bytes",
                "Type": "bigint",
              },
              {
                "Name": "time_taken",
                "Type": "float",
              },
              {
                "Name": "x_forwarded_for",
                "Type": "string",
              },
              {
                "Name": "ssl_protocol",
                "Type": "string",
              },
              {
                "Name": "ssl_cipher",
                "Type": "string",
              },
              {
                "Name": "x_edge_response_result_type",
                "Type": "string",
              },
              {
                "Name": "cs_protocol_version",
                "Type": "string",
              },
              {
                "Name": "fle_status",
                "Type": "string",
              },
              {
                "Name": "fle_encrypted_fields",
                "Type": "int",
              },
              {
                "Name": "c_port",
                "Type": "int",
              },
              {
                "Name": "time_to_first_byte",
                "Type": "float",
              },
              {
                "Name": "x_edge_detailed_result_type",
                "Type": "string",
              },
              {
                "Name": "sc_content_type",
                "Type": "string",
              },
              {
                "Name": "sc_content_len",
                "Type": "bigint",
              },
              {
                "Name": "sc_range_start",
                "Type": "bigint",
              },
              {
                "Name": "sc_range_end",
                "Type": "bigint",
              },
            ],
            "InputFormat": "org.apache.hadoop.mapred.TextInputFormat",
            "Location": {
              "Fn::Join": [
                "",
                [
                  "s3://",
                  {
                    "Ref": "LogBucketCC3B17E8",
                  },
                  "/path/to/location/",
                ],
              ],
            },
            "OutputFormat": "org.apache.hadoop.hive.ql.io.HiveIgnoreKeyTextOutputFormat",
            "SerdeInfo": {
              "Parameters": {
                "field.delim": "	",
              },
              "SerializationLibrary": "org.apache.hadoop.hive.serde2.lazy.LazySimpleSerDe",
            },
          },
          "TableType": "EXTERNAL_TABLE",
        },
      },
      "Type": "AWS::Glue::Table",
    },
    "LogBucketCC3B17E8": {
      "DeletionPolicy": "Retain",
      "Type": "AWS::S3::Bucket",
      "UpdateReplacePolicy": "Retain",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
